### PR TITLE
fix: consistency in display reserved_stock checkbox on Sales Order Item according global settings and item.is_stock_item

### PR DESCRIPTION
**Pull request by FHenry**
> Please provide enough information so that others can review your pull request:

1 : Global setting consistency in Sales Order Item grid table
If Reserved Stock In Stock Settings is OFF, the reserve stock column should not be available in column choice on Sales Order Item grid table and reserve_stock must be set to 0 on Sales Order Item data
It give to user headache to understand what is the meaning of reserve stock column

2 : Maintain Stock consistency in Sales Order Item grid table
If Reserved Stock In Stock Settings is ON
If an Item is **not** Maintain Stock there is no need to display Reserve Stock checkbox field in Sales Order Item detail form.
It give to user headache to understand what is the meaning of reserve stock checkbox for a non Maintain Stock item
The value must be defaulted to 0 in this case.

> Explain the **details** for making this change. What existing problem does the pull request solve?

Item "Item 1" is Maintain Stock
Item "no stock" is not Maintain Stock

Actually :
Add Item "Item 1" to a Sales Order The Reserve stock checkbox is in detail Sales Order Item Form => OK
Add Item "no stock" to a Sales Order The Reserve stock checkbox is in detail Sales Order Item Form and defaulted to 1 => KO

With this PR :
Add Item "Item 1" to a Sales Order The Reserve stock checkbox is in detail Sales Order Item Form => OK
Add Item "no stock" to a Sales Order The Reserve stock checkbox is **not** in detail Sales Order Item Form and defaulted to 0 => OK

> Screenshots/GIFs

Before :

https://github.com/frappe/erpnext/assets/1050053/4f2b44fd-05c7-489f-92a2-e64713c2c046

After :
Stock Setting Reserved Stock OFF

https://github.com/frappe/erpnext/assets/1050053/bcdf7d20-c7c2-4002-9a82-60ba74bb0f66

Stock Setting Reserved Stock ON with non Maintain Stock item

https://github.com/frappe/erpnext/assets/1050053/ba9e1266-ccfa-4c91-aad4-4928275dfda9

> no-doc

### CODE REVIEW

File: erpnext/selling/doctype/sales_order_item/sales_order_item.json [hunk 1]
```diff
@@ -10,6 +10,7 @@
   "item_code",
   "customer_item_code",
   "ensure_delivery_based_on_produced_serial_no",
+  "is_stock_item",
```
Comment (dj12djdjs) [inline erpnext/selling/doctype/sales_order_item/sales_order_item.json:13]: Is this required? It looks like you're checking the cached value from item during validation?

**Comment by FHenry** [comment]
Note to reviewer : the failed test MariaDB (Test 3) is not related to this PR.

But If you want to suggest me different implementation  to get is_stock_item from item, that could be more Frappe Ways , than adding a field in Sales Order Item : Is Stock Item (hidden and defaulted with item.is_stock_item)
like use async/await in erpnext/buying/doctype/purchase_order/purchase_order.js 
`frappe.ui.form.on("Purchase Order Item", { item_code: ...`

I will be happy to work on it

**Comment by dj12djdjs** [comment]
@FHenry I mean I don't see the `is_stock_item` field you created used anywhere. 

I think it can be deleted, or am I totally missing something here?

**Comment by FHenry** [comment]
> @FHenry I mean I don't see the `is_stock_item` field you created used anywhere.
> 
> I think it can be deleted, or am I totally missing something here?

@dj12djdjs 
I use it in the json
  "depends_on": "eval:doc.is_stock_item",
of the field rerserve_stock

Like that when an Item is fetch 
Sales Order Item.is_stock_item 
 is populated with 
item.is_stock_item 
and then reserve_stock is displayed or not

**Comment by FHenry** [comment]
@dj12djdjs did you have time to check this PR ?

File: .editorconfig [hunk 2]
```diff
@@ -12,3 +12,6 @@ charset = utf-8
 [{*.py,*.js}]
 indent_style = tab
 indent_size = 4
+
+[*.json]
+insert_final_newline = false
```
Comment (s-aga-r) [inline .editorconfig]: Undo these changes.

File: erpnext/selling/doctype/sales_order/sales_order.js [hunk 3]
```diff
@@ -94,6 +94,15 @@ frappe.ui.form.on("Sales Order", {
...
 						frm.set_df_property("reserve_stock", "hidden", 1);
+						frm.fields_dict.items.grid.update_docfield_property(
+								'reserve_stock', 'hidden', 1
+						);
+						frm.fields_dict.items.grid.update_docfield_property(
+								'reserve_stock', 'default', 0
+						);
+						frm.fields_dict.items.grid.update_docfield_property(
+								'reserve_stock', 'read_only', 1
+						);
```
Comment (s-aga-r) [inline erpnext/selling/doctype/sales_order/sales_order.js]: Fix spacing.

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 4]
```diff
@@ -660,6 +661,15 @@ def validate_serial_no_based_delivery(self):
...
 				)
 
+	def validate_reserved_stock(self):
+		"""Clean reserved stock flag for not stocked Item"""
+
+		for item in self.items:
+			if not cint(
+				frappe.get_cached_value("Item", item.item_code, "is_stock_item")
+			) or not frappe.db.get_single_value("Stock Settings", "enable_stock_reservation"):
+				item.reserve_stock = 0
```
Comment (s-aga-r) [inline erpnext/selling/doctype/sales_order/sales_order.py]: get `enable_stock_reservation` from DB once and use it in the iteration.

Why we are fetching `is_stock_item` from Item master as you have added a field is_stock_item in the Sales Order Item itself?

**Review by s-aga-r — CHANGES_REQUESTED** [review]

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 5]
```diff
@@ -660,6 +661,15 @@ def validate_serial_no_based_delivery(self):
...
 
+	def validate_reserved_stock(self):
+		"""Clean reserved stock flag for not stocked Item"""
+
+		for item in self.items:
+			if not cint(item.is_stock_item) or not frappe.db.get_single_value(
+				"Stock Settings", "enable_stock_reservation"
+			):
+				item.reserve_stock = 0
+
```
Comment (s-aga-r) [inline erpnext/selling/doctype/sales_order/sales_order.py]: get `enable_stock_reservation` from DB once and use it in the iteration.

**Review by s-aga-r — CHANGES_REQUESTED** [review]

**Review by s-aga-r — APPROVED** [review]

**Comment by s-aga-r** [comment]
Unrelated failing tests.

**Pull request merged**